Redefine (co)limits in terms of kan extensions (#186)
plt-amy/1lab: src/Cat/Diagram/Colimit/Base.lagda.md

that we define a colimits of a diagram $F$ as [left kan extensions]
instead of [right kan extensions]. This gives us the expected
"mapping out" universal property, as opposed to the "mapping in" property
associated to limits.

[left kan extension]: Cat.Functor.Kan.Left.html
[right kan extension]: Cat.Functor.Kan.Right.html

Note that approach to colimits is not what normally presented in
introductory material. Instead, most books opt to define colimits
via [cocones], as they are less abstract, though harder to work with
in the long run.

[cocones]: Cat.Diagram.Colimit.Cocone.html

<!--
```agda
private variable
  o₁ o₂ o₃ h₁ h₂ h₃ : Level
```
-->

```agda
module _ {J : Precategory o₁ h₁} {C : Precategory o₂ h₂} (Diagram : Functor J C) where
  private
    module C = Precategory C

  cocone→unit : ∀ {x : C.Ob} → (Diagram => Const x) → Diagram => const! x F∘ !F
  unquoteDef cocone→unit = define-coherence cocone→unit

  is-colimit : (x : C.Ob) → Diagram => Const x → Type _
  is-colimit x cocone =
    is-lan !F Diagram (const! x) (cocone→unit cocone)

  Colimit : Type _
  Colimit = Lan !F Diagram
```

## Concretely

> The above definition is very concise, and has the benefit of being
> abstract: this means that we can directly re-use definitions and
> theorems for kan extensions for colimits. However, this also means
> that the definition is abstract: it makes working with colimits
> _in general_ easier, at the cost of making _specific_ colimits more
> difficult, as the data we actually care about has been obfuscated.
> 
> For instance, it's quite hard to actually show that a specific object
> is a colimit of aa diagram, as you have to wrap all the relevant data
> in a few layers of abstraction. To work around this, we provide an
> auxiliary record type, `make-is-colimit`{.Agda}, which computes the
> appropriate left extension.

Reviewer: ```suggestion
As mentioned, our definition is very abstract, meaning we can directly
re-use definitions and theorems about Kan extensions in the setting of
colimits. The trade-off is that while working with colimits _in general_
is easier, working with _specific_ colimits becomes more difficult, as
the data we actually care about has been obfuscated.

One particularly egregious failure is... actually constructing colimits.
The definition in terms of `Lan`{.Agda} hides the concrete data behind a
few abstractions, which would be very tedious to write out each time. To
work around this, we provide an auxiliary record type,
`make-is-colimit`{.Agda}, as an intermediate step in constructing left
extensions.
```

<!--
```agda
module _ {J : Precategory o₁ h₁} {C : Precategory o₂ h₂}
  where
  private
    module J = Precategory J
    module C = Cat.Reasoning C

  record make-is-colimit
    (Diagram : Functor J C) (coapex : C.Ob)
    : Type (o₁ ⊔ h₁ ⊔ o₂ ⊔ h₂)
    where
    no-eta-equality
    open Functor Diagram
```

> -->A

Reviewer: A

> First, we require morphisms from the every value to the diagram to

Reviewer: ```suggestion
First, we require morphisms from the every value of the diagram to
```

> the coapex; we call this family $\phi$. Moreover, if $f : x \to y$ is

Reviewer: ```suggestion
the coapex; taken as a family, we call it $\phi$. Moreover, if $f : x \to y$ is
```

> a morphism in the "shape" category $\cJ$, then $\psi y \circ Ff = \psi x$,
> IE: the $\psi$ family is actually natural.

Reviewer: ```suggestion
a morphism in the "shape" category $\cJ$, we require $\psi y \circ Ff = \psi x$,
which encodes the relevant naturality.
```

```agda
    field
      ψ : (j : J.Ob) → C.Hom (F₀ j) coapex
      commutes : ∀ {x y} (f : J.Hom x y) → ψ y C.∘ F₁ f ≡ ψ x
```

The rest of the data ensures that $\psi$ is the universal family

> of maps iwth this property; if $\epsilon_j : Fj \to x$ is another natural

Reviewer: ```suggestion
of maps with this property; if $\eps_j : Fj \to x$ is another natural
```

> family, then each $\epsilon_j$ factors through the coapex by a _unique_

Reviewer: ```suggestion
family, then each $\eps_j$ factors through the coapex by a _unique_
```

    colim .σ {M = M} α .η _ =
      universal (α .η) (λ f → α .is-natural _ _ f ∙ C.eliml (M .F-id))
    colim .σ {M = M} α .is-natural _ _ _ =
       C.idr _ ∙ C.introl (M .F-id)
    colim .σ-comm {α = α} = Nat-path λ j →
      factors (α .η) _
    colim .σ-uniq {α = α} {σ′ = σ′} p = Nat-path λ _ →
      sym $ unique (α .η) _ (σ′ .η _) (λ j → sym (p ηₚ j))
```

<!--
```agda
  to-is-colimitp
    : ∀ {D : Functor J C} {K : Functor ⊤Cat C} {eta : D => K F∘ !F}
    → (mk : make-is-colimit D (Functor.F₀ K tt))
    → (∀ {j} → to-cocone mk .η j ≡ eta .η j)
    → is-lan !F D K eta
  to-is-colimitp {D} {K} {eta} mkcolim p = colim where
    open make-is-colimit mkcolim
    open is-lan
    open Functor
    open _=>_

    colim : is-lan !F D K eta
    colim .σ {M = M} β .η _ =
      universal (β .η) (λ f → β .is-natural _ _ f ∙ C.eliml (M .F-id))
    colim .σ {M = M} β .is-natural _ _ _ =
      C.elimr (K .F-id) ∙ C.introl (M .F-id)
    colim .σ-comm {α = α} = Nat-path λ j →
      ap (_ C.∘_) (sym p) ∙ factors (α .η) _
    colim .σ-uniq {α = α} {σ′ = σ′} q = Nat-path λ _ →
      sym $ unique (α .η) _ (σ′ .η tt) λ j →
        ap (_ C.∘_) p ∙ sym (q ηₚ j)
```
-->

> We also want to be able to use the interface of `make-is-colimit`{.Agda}
> when we have our hands on a colimit. To do this, we provide a function
> for *un*making a colimit.

Reviewer: ```suggestion
The concrete interface of `make-is-colimit`{.Agda} is also handy for
_consuming_ specific colimits. To enable this use case, we provide a
function which **un**makes a colimit.
```

```

<!--
```agda
  to-colimit
    : ∀ {D : Functor J C} {K : Functor ⊤Cat C} {eta : D => K F∘ !F}
    → is-lan !F D K eta
    → Colimit D
  to-colimit c .Lan.Ext = _
  to-colimit c .Lan.eta = _
  to-colimit c .Lan.has-lan = c
```
-->

<!--
```agda
module is-colimit
  {J : Precategory o₁ h₁} {C : Precategory o₂ h₂}
  {D : Functor J C} {F : Functor ⊤Cat C} {eta : D => F F∘ !F}
  (t : is-lan !F D F eta)
  where

  open make-is-colimit (unmake-colimit {F = F} t) public
```
-->

We also provide a similar interface for the bundled form of colimits.

```agda
module Colimit
  {J : Precategory o₁ h₁} {C : Precategory o₂ h₂} {D : Functor J C} (L : Colimit D)
  where
```

<!--
```agda
  private
    import Cat.Reasoning J as J
    import Cat.Reasoning C as C
    module Diagram = Functor D
    open Lan L
    open Functor
    open _=>_
```
-->

The coapex of the colimit can be obtained by applying the extension functor
to the single object of `⊤Cat`{.Agda}.

```agda
  coapex : C.Ob
  coapex = Ext .F₀ tt
```

Furthermore, we can show that the apex is the colimit, in the sense of
`is-colimit`{.Agda}, of the diagram. You'd think this is immediate, but

> unfortunately proof assistants: `is-colimit`{.Agda} asks for _the_

Reviewer: ```suggestion
unfortunately, proof assistants: `is-colimit`{.Agda} asks for _the_
```

Unfortunately, proof assistants always have a comma!

constant functor functor $\{*\} \to \cC$ with value `coapex` to be a Kan
extension, but `Colimit`{.Agda}, being an instance of `Lan`{.Agda},
packages an _arbitrary_ functor $\{*\} \to \cC$.

Since Agda does not compare functors for $\eta$-equality, we have to
shuffle our data around manually. Fortunately, this isn't a very long
computation.

```agda
  cocone : D => Const coapex
  cocone .η = eta .η
  cocone .is-natural x y f =
    eta .is-natural x y f ∙ ap (C._∘ _) (Ext .F-id)

  has-colimit : is-colimit D coapex cocone
  has-colimit .is-lan.σ α .η = σ α .η
  has-colimit .is-lan.σ α .is-natural x y f =
    ap (_ C.∘_) (sym (Ext .F-id)) ∙ σ α .is-natural tt tt tt
  has-colimit .is-lan.σ-comm =
    Nat-path (λ _ → σ-comm ηₚ _)
  has-colimit .is-lan.σ-uniq {M = M} {σ′ = σ′} p =
    Nat-path (λ _ → σ-uniq {σ′ = nt} (Nat-path (λ j → p ηₚ j)) ηₚ _)
    where
      nt : Ext => M
      nt .η = σ′ .η
      nt .is-natural x y f = ap (_ C.∘_) (Ext .F-id) ∙ σ′ .is-natural x y f

  open is-colimit has-colimit public
```

> # Uniqueness

Reviewer: This is an instance of uniqueness of Lan, nay? Can't we reuse that? `Cat.Functor.Kan.Unique`

Reviewer: All of this duplication really isn't making our case that working with the abstract definition is better..! We should have all of these lemmas for arbitrary `is-lan`, in a module about extensions.

Author: Done!

```
-->

```agda
  colimits→inversesp
    : ∀ {f : C.Hom x y} {g : C.Hom y x}
    → (∀ {j : J.Ob} → f C.∘ Cx.ψ j ≡ Cy.ψ j)
    → (∀ {j : J.Ob} → g C.∘ Cy.ψ j ≡ Cx.ψ j)
    → C.Inverses f g
  colimits→inversesp f-factor g-factor =
    C.make-inverses
      (Cy.unique₂ Cy.ψ Cy.commutes (λ j → C.pullr g-factor ∙ f-factor) λ _ → C.idl _)
      (Cx.unique₂ Cx.ψ Cx.commutes (λ j → C.pullr f-factor ∙ g-factor) λ _ → C.idl _)

  colimits→invertiblep
    : ∀ {f : C.Hom x y}
    → (∀ {j : J.Ob} → f C.∘ Cx.ψ j ≡ Cy.ψ j)
    → C.is-invertible f
  colimits→invertiblep f-factor =
    C.inverses→invertible $
    colimits→inversesp f-factor (Cy.factors Cx.ψ Cx.commutes)

  colimits→inverses
    : C.Inverses (Cx.universal Cy.ψ Cy.commutes) (Cy.universal Cx.ψ Cx.commutes)
  colimits→inverses =
    colimits→inversesp (Cx.factors Cy.ψ Cy.commutes) (Cy.factors Cx.ψ Cx.commutes)

  colimits→invertible
    : C.is-invertible (Cx.universal Cy.ψ Cy.commutes)
  colimits→invertible = colimits→invertiblep (Cx.factors Cy.ψ Cy.commutes)

  colimits-unique : x C.≅ y
  colimits-unique = C.invertible→iso _ colimits→invertible
```

Furthermore, if the universal map is invertible, then that means its

> domain is _also_ a colimit of the diagram.

Reviewer: This should also be proved in more generality

<!--
```agda
module _ {o₁ h₁ o₂ h₂ : _} {J : Precategory o₁ h₁} {C : Precategory o₂ h₂}
         {D : Functor J C} {K : Functor ⊤Cat C}
         {etay : D => Const (Functor.F₀ K tt)}
         (Cy : is-colimit D (Functor.F₀ K tt) etay)
       where
  private
    module J = Precategory J
    module C = Cat.Reasoning C
    module D = Functor D
    open is-ran
    open Functor
    open _=>_

    module Cy = is-colimit Cy

  family→cocone
    : ∀ {x}
    → (eps : ∀ j → C.Hom (D.₀ j) x)
    → (∀ {x y} (f : J.Hom x y) → eps y C.∘ D.₁ f ≡ eps x)
    → D => Const x
  family→cocone eta p .η = eta
  family→cocone eta p .is-natural _ _ _ = p _ ∙ sym (C.idl _)
```
-->

```agda

>   is-invertible→is-colimitp

Reviewer: This should also be proved in more generality

    : ∀ {K' : Functor ⊤Cat C} {eta : D => K' F∘ !F}
    → (eps : ∀ j → C.Hom (D.₀ j) (K' .F₀ tt))
    → (p : ∀ {x y} (f : J.Hom x y) → eps y C.∘ D.₁ f ≡ eps x)
    → (∀ {j} → eps j ≡ eta .η j)
    → C.is-invertible (Cy.universal eps p)
    → is-lan !F D K' eta
  is-invertible→is-colimitp {K' = K'} eps p q invert =
    to-is-colimitp colim q
    where
      open C.is-invertible invert
      open make-is-colimit

      colim : make-is-colimit D (K' .F₀ tt)
      colim .ψ = eps
      colim .commutes = p
      colim .universal tau q = Cy.universal tau q C.∘ inv
      colim .factors tau q =
        sym (C.refl⟩∘⟨ Cy.factors eps p)
        ·· C.cancel-inner invr
        ·· Cy.factors tau q
      colim .unique tau q other r =
        C.insertr invl
        ∙ (Cy.unique _ _ _ (λ j → C.pullr (Cy.factors eps p) ∙ r j) C.⟩∘⟨refl)
```

Another useful fact is that if $C$ is a colimit of some diagram $Dia$,
and $Dia$ is naturally isomorphic to some other diagram $Dia'$, then the
coapex of $C$ is also a colimit of $Dia'$.

```agda

>   natural-iso→is-colimitp

Reviewer: This should be also proved in more generality. Have a function that allows varying every argument of `is-lan` by a natural isomorphsim

<!--
```agda
module _ {J : Precategory o₁ h₁} {C : Precategory o₂ h₂} {D : Precategory o₃ h₃}
         (F : Functor C D) {Diagram : Functor J C} where
  private
    module D = Precategory D
    module C = Precategory C
    module J = Precategory J
    module F = Func F
```
-->

```agda
  preserves-colimit
     : ∀ {K : Functor ⊤Cat C} {eta : Diagram => K F∘ !F}
     → is-lan !F Diagram K eta
     → Type _
  preserves-colimit colim = preserves-lan F colim

  reflects-colimit
    : ∀ {K : Functor ⊤Cat C} {eps : Diagram => K F∘ !F}
    → is-lan !F (F F∘ Diagram) (F F∘ K) (nat-assoc-to (F ▸ eps))
    → Type _
  reflects-colimit lan = reflects-lan F lan

--   record creates-colimit : Type (o₁ ⊔ h₁ ⊔ o₂ ⊔ h₂ ⊔ o₃ ⊔ h₃) where
--     field
--       preserves-colimit : Preserves-colimit
--       reflects-colimit : Reflects-colimit
```

## Cocontinuity

```agda
is-cocontinuous
  : ∀ {oshape hshape}
      {C : Precategory o₁ h₁}
      {D : Precategory o₂ h₂}
  → Functor C D → Type _
```

A cocontinuous functor is one that --- for every shape of diagram `J`,
and every diagram `diagram`{.Agda} of shape `J` in `C` --- preserves the

> colimit for that diagram.

Reviewer: I'm not a big fan of the em dashes here
